apps/browser/src/vault/popup/services/vault-popup-autofill.service.ts

```diff
@@ -0,0 +1,274 @@
+   *
+   * The page details collection process has no definite 'end', so we must keep a subscription alive
+   * to ensure that the most up-to-date page details are always available for autofill.
+   *
+   * This method should be called early whenever autofill may be needed, such as when the popup is opened.
+   */
+  startCollectingPageDetails() {
+    if (this._pageDetailsSubscription && !this._pageDetailsSubscription.closed) {
+      return;
+    }
+    this._pageDetailsSubscription = this._currentPageDetails$.subscribe();
+  }
+
+  /**
+   * Stop the subscription to collect the page details from the current tab.
+   *
+   * This can be called when navigating away from the vault or wherever autofill is no longer needed.
+   */
+  stopCollectingPageDetails() {
+    if (this._pageDetailsSubscription && !this._pageDetailsSubscription.closed) {
+      this._pageDetailsSubscription.unsubscribe();
+      this._pageDetailsSubscription = null;
+    }
+  }
+
+  /**
+   * Helper to log a warning if there is an attempt to autofill without an active page details subscription.
+   * @private
+   */
+  private _warnIfMissingPageDetailsSub() {
+    if (!this._pageDetailsSubscription || this._pageDetailsSubscription.closed) {
+      this.logService.warning(
+        "Page details subscription is not active. Page details may not be available in time for autofill.",
+      );
+    }
+  }
+
+  private async _internalDoAutofill(
+    cipher: CipherView,
+    tab: chrome.tabs.Tab,
+    pageDetails: PageDetail[],
+  ): Promise<boolean> {
+    if (
+      cipher.reprompt !== CipherRepromptType.None &&
+      !(await this.passwordRepromptService.showPasswordPrompt())
+    ) {
+      return false;
+    }
+
+    if (tab == null || pageDetails.length === 0) {
+      this.toastService.showToast({
+        variant: "error",
+        title: null,
+        message: this.i18nService.t("autofillError"),
+      });
+      return false;
+    }
+
+    try {
+      const totpCode = await this.autofillService.doAutoFill({
+        tab,
+        cipher,
+        pageDetails,
+        doc: window.document,
+        fillNewPassword: true,
+        allowTotpAutofill: true,
+      });
+
+      if (totpCode != null) {
+        this.platformUtilService.copyToClipboard(totpCode, { window: window });
+      }
+    } catch {
+      this.toastService.showToast({
+        variant: "error",
+        title: null,
+        message: this.i18nService.t("autofillError"),
+      });
+      return false;
+    }
+
+    return true;
+  }
+
+  private _closePopup() {
+    if (BrowserPopupUtils.inPopup(window)) {
+      if (this.platformUtilService.isFirefox() || this.platformUtilService.isSafari()) {
+        BrowserApi.closePopup(window);
+      } else {
+        // Slight delay to fix bug in Chromium browsers where popup closes without copying totp to clipboard
+        setTimeout(() => BrowserApi.closePopup(window), 50);
+      }
+    }
+  }
```

Comment:
```suggestion
  private _closePopup() {
    if (!BrowserPopupUtils.inPopup(window)) {
      return;
    }

    if (this.platformUtilService.isFirefox() || this.platformUtilService.isSafari()) {
        BrowserApi.closePopup(window);
        return;
    } 

    // Slight delay to fix bug in Chromium browsers where popup closes without copying totp to clipboard
    setTimeout(() => BrowserApi.closePopup(window), 50);
  }
```

🎨 Nit - This might be more readable if we avoid nesting of conditionals with early returns. Not a blocker, just a suggestion.

Reply:
I like it, done! [0f1b54a](https://github.com/bitwarden/clients/pull/9616/commits/0f1b54a866186bc75a921c982add26ea71fe4692)